src/GraphQL/Internal/Name.hs

```diff
@@ -17,13 +19,72 @@ module GraphQL.Internal.Name
 
 import Protolude
 
+import qualified Data.Aeson as Aeson
 import GHC.TypeLits (Symbol, KnownSymbol, symbolVal)
-import GraphQL.Internal.Syntax.AST
-  ( Name(..)
-  , NameError(..)
-  , unsafeMakeName
-  , makeName
-  )
+import Data.Char (isDigit)
+import Data.Text as T (Text)
+import qualified Data.Attoparsec.Text as A
+import Test.QuickCheck (Arbitrary(..), elements, listOf)
+import Data.String (IsString(..))
+import Data.Text as T (Text, append, empty)
+
+import GraphQL.Internal.Syntax.Tokens (tok)
+
+-- * Name
+
+-- | A name in GraphQL.
+--
+-- https://facebook.github.io/graphql/#sec-Names
+newtype Name = Name { unName :: T.Text } deriving (Eq, Ord, Show)
+
+instance Monoid Name where
```

Comment:
Why is this instance useful? Would you mind adding a comment that explains why we have this instance?

Reply:
### What it's doing
If Name is not a monoid, then `optempty` cannot operate on nameParser (because without it, it only produces a Name or a NameError). As a monoid, optempty returns a Name "SomeValidName" or Name mempty. This allows the following:
``` 
node :: Parser AST.Node
  node = AST.Node <$> optempty nameParser
                  <*> optempty variableDefinitions
                  <*> optempty directives
                  <*> selectionSet
```
This makes graphql-api`s implementation more forgiving than the spec, without any added ambiguity (because of validation that follows in our validation layer). 
### Justification & Discussion
According to the spec,  [#operation-name](http://graphql.org/learn/queries/#operation-name): 
>Up until now, we have been using a shorthand syntax where we omit both the query keyword and the query name, but in production apps it's useful to use these to make our code less ambiguous. You'll need these optional parts to a GraphQL operation if you want to execute something other than a query or pass dynamic variables.

It says, that BOTH *query keyword* AND *query name* MUST be added to *mutation*, *subscription*, or a *query* that passes *dynamic variables*. Unfortunately, it is not this strongly written however, and because *query name* (operation name) is practically useless, client implementations I've encountered don't follow it strictly.

The utility of *operation names* is in the contrived case where multiple operations are sent in a single document (_dead load because there is absolutely no use, now and possibly ever for, these extra superfluous operations in the document: look [here](https://github.com/jamesmacaulay/elm-graphql/issues/30#issuecomment-353277368) for some of the justifications I've encountered_). Note that I ultimately arguing to USE operation names in the referenced comment.

The spec is NOT strongly written enough to preclude the possibility of client implementations sending something like:
```
query {greeting(who: "Tim")}
```
which we currently fail on, but could theoretically turn into an AST.AnonymousQuery by changing:
```
operationDefinition :: Parser AST.OperationDefinition$
  operationDefinition =
        AST.Query    <$ tok "query"    <*> node
    <|> AST.Mutation <$ tok "mutation" <*> node
    <|> (AST.AnonymousQuery <$> selectionSet)
    <?> "operationDefinition error!"
```
into something like this (untested, not 100% "query" is consumed on failure of AST.Query match):
```
operationDefinition :: Parser AST.OperationDefinition$
  operationDefinition =
        AST.Query    <$ tok "query"    <*> node
    <|> (AST.AnonymousQuery <$> selectionSet)
    <|> AST.Mutation <$ tok "mutation" <*> node
    <|> (AST.AnonymousQuery <$> selectionSet)
    <?> "operationDefinition error!"
```
OR we can do what I have done in this PR, which is make the `Name` optional, and _have a nice day_ :)  In addition, although it's out of spec, we handle the following case as well:
 ```
query ($name : String) { greeting(who: $name) }
``` 
I.e. A request explicitly coded as a query, but without a name, and takes variables (this is not possible to misinterpret as anything other than a violation of the spec), but there is no danger in doing so because the validator will pick up ambiguity in the case of multiple operations AND no valid JSON operation name.